<!DOCTYPE html>
<html>

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>HGCC_StartGuide_2</title>


<style type="text/css">
body {
  font-family: Helvetica, arial, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  padding-top: 10px;
  padding-bottom: 10px;
  background-color: white;
  padding: 30px; }

body > *:first-child {
  margin-top: 0 !important; }
body > *:last-child {
  margin-bottom: 0 !important; }

a {
  color: #4183C4; }
a.absent {
  color: #cc0000; }
a.anchor {
  display: block;
  padding-left: 30px;
  margin-left: -30px;
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0; }

h1, h2, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
  cursor: text;
  position: relative; }

h1:hover a.anchor, h2:hover a.anchor, h3:hover a.anchor, h4:hover a.anchor, h5:hover a.anchor, h6:hover a.anchor {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA09pVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoMTMuMCAyMDEyMDMwNS5tLjQxNSAyMDEyLzAzLzA1OjIxOjAwOjAwKSAgKE1hY2ludG9zaCkiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OUM2NjlDQjI4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OUM2NjlDQjM4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo5QzY2OUNCMDg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo5QzY2OUNCMTg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PsQhXeAAAABfSURBVHjaYvz//z8DJYCRUgMYQAbAMBQIAvEqkBQWXI6sHqwHiwG70TTBxGaiWwjCTGgOUgJiF1J8wMRAIUA34B4Q76HUBelAfJYSA0CuMIEaRP8wGIkGMA54bgQIMACAmkXJi0hKJQAAAABJRU5ErkJggg==) no-repeat 10px center;
  text-decoration: none; }

h1 tt, h1 code {
  font-size: inherit; }

h2 tt, h2 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h4 tt, h4 code {
  font-size: inherit; }

h5 tt, h5 code {
  font-size: inherit; }

h6 tt, h6 code {
  font-size: inherit; }

h1 {
  font-size: 28px;
  color: black; }

h2 {
  font-size: 24px;
  border-bottom: 1px solid #cccccc;
  color: black; }

h3 {
  font-size: 18px; }

h4 {
  font-size: 16px; }

h5 {
  font-size: 14px; }

h6 {
  color: #777777;
  font-size: 14px; }

p, blockquote, ul, ol, dl, li, table, pre {
  margin: 15px 0; }

hr {
  background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAYAAAAECAYAAACtBE5DAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBNYWNpbnRvc2giIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OENDRjNBN0E2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OENDRjNBN0I2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo4Q0NGM0E3ODY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo4Q0NGM0E3OTY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PqqezsUAAAAfSURBVHjaYmRABcYwBiM2QSA4y4hNEKYDQxAEAAIMAHNGAzhkPOlYAAAAAElFTkSuQmCC) repeat-x 0 0;
  border: 0 none;
  color: #cccccc;
  height: 4px;
  padding: 0;
}

body > h2:first-child {
  margin-top: 0;
  padding-top: 0; }
body > h1:first-child {
  margin-top: 0;
  padding-top: 0; }
  body > h1:first-child + h2 {
    margin-top: 0;
    padding-top: 0; }
body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
  margin-top: 0;
  padding-top: 0; }

a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0; }

h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
  margin-top: 0; }

li p.first {
  display: inline-block; }
li {
  margin: 0; }
ul, ol {
  padding-left: 30px; }

ul :first-child, ol :first-child {
  margin-top: 0; }

dl {
  padding: 0; }
  dl dt {
    font-size: 14px;
    font-weight: bold;
    font-style: italic;
    padding: 0;
    margin: 15px 0 5px; }
    dl dt:first-child {
      padding: 0; }
    dl dt > :first-child {
      margin-top: 0; }
    dl dt > :last-child {
      margin-bottom: 0; }
  dl dd {
    margin: 0 0 15px;
    padding: 0 15px; }
    dl dd > :first-child {
      margin-top: 0; }
    dl dd > :last-child {
      margin-bottom: 0; }

blockquote {
  border-left: 4px solid #dddddd;
  padding: 0 15px;
  color: #777777; }
  blockquote > :first-child {
    margin-top: 0; }
  blockquote > :last-child {
    margin-bottom: 0; }

table {
  padding: 0;border-collapse: collapse; }
  table tr {
    border-top: 1px solid #cccccc;
    background-color: white;
    margin: 0;
    padding: 0; }
    table tr:nth-child(2n) {
      background-color: #f8f8f8; }
    table tr th {
      font-weight: bold;
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr td {
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr th :first-child, table tr td :first-child {
      margin-top: 0; }
    table tr th :last-child, table tr td :last-child {
      margin-bottom: 0; }

img {
  max-width: 100%; }

span.frame {
  display: block;
  overflow: hidden; }
  span.frame > span {
    border: 1px solid #dddddd;
    display: block;
    float: left;
    overflow: hidden;
    margin: 13px 0 0;
    padding: 7px;
    width: auto; }
  span.frame span img {
    display: block;
    float: left; }
  span.frame span span {
    clear: both;
    color: #333333;
    display: block;
    padding: 5px 0 0; }
span.align-center {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-center > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: center; }
  span.align-center span img {
    margin: 0 auto;
    text-align: center; }
span.align-right {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-right > span {
    display: block;
    overflow: hidden;
    margin: 13px 0 0;
    text-align: right; }
  span.align-right span img {
    margin: 0;
    text-align: right; }
span.float-left {
  display: block;
  margin-right: 13px;
  overflow: hidden;
  float: left; }
  span.float-left span {
    margin: 13px 0 0; }
span.float-right {
  display: block;
  margin-left: 13px;
  overflow: hidden;
  float: right; }
  span.float-right > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: right; }

code, tt {
  margin: 0 2px;
  padding: 0 5px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px; }

pre code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent; }

.highlight pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }

pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }
  pre code, pre tt {
    background-color: transparent;
    border: none; }

sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}

kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb
}

* {
	-webkit-print-color-adjust: exact;
}
@media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:0 auto;
    }
}
@media print {
	table, pre {
		page-break-inside: avoid;
	}
	pre {
		word-wrap: break-word;
	}
  body {
    padding: 2cm; 
  }
}
</style>


</head>

<body>

<h1 id="toc_0">Run Computation Tools and Submit Jobs on HGCC</h1>

<h2 id="toc_1">Test commands in an interactive session on computation node</h2>

<ol>
<li>Open an interactive bash session on a computation node by <code>srun -N 1 -n 1 --pty bash -i</code>.</li>
<li>Load required software modules by <code>spack load [software]</code>.</li>
<li>Mount remote HGCC directory to your local computer.</li>
<li>Write scripts using local text editors like <strong><code>Visual Studio Code</code>, <code>Sublime</code></strong>.</li>
<li>Copy and paste to the interactive bash session.</li>
</ol>

<h3 id="toc_2">R</h3>

<ul>
<li>Load R module by <code>spack load r@4.4.0</code>.</li>
<li>Type <strong><code>R</code></strong> in the interactive session to initiate an R session.</li>
<li><strong>Copy R commands</strong> from your R script and <strong>Paste into the R session</strong>.</li>
<li>Simple parallel computation can be enabled by using R packages <strong><code>parallel</code></strong> and <strong><code>foreach</code></strong>, see <strong><a href="https://nceas.github.io/oss-lessons/parallel-computing-in-r/parallel-computing-in-r.html">Quick Instructions of Parallel Computation</a></strong>.</li>
<li>Tips

<ul>
<li>Provide full directory to the data files you want to read or write or plot.</li>
<li>Save figures as pdf into the disk to view your plots by R function <code>ggplot2::ggsave(&quot;temp.pdf&quot;)</code>.</li>
</ul></li>
</ul>

<h3 id="toc_3">Python/Conda</h3>

<ul>
<li><strong>Python 3.11.7</strong> is installed on the cluster.</li>
<li>Interactive <strong>python</strong> session:

<ol>
<li>Type <strong><code>python</code></strong> in the interactive session to initiate a python session.</li>
<li><strong>Copy python commands</strong> from your python script and <strong>Paste into the python session</strong>.</li>
</ol></li>
<li>Use <strong>Virtual Environment</strong> for access right to install python libraries.

<ol>
<li>Load miniconda by <code>spack load miniconda3@24.3.0</code>.</li>
<li>Run <code>conda init; source ~/.bashrc;</code> for once for using <code>conda</code>.</li>
<li><strong>Create</strong> a new virtual environment with name <strong>myenv</strong>: <strong><code>conda create -n &quot;myenv&quot;</code></strong>.</li>
<li><strong>List</strong> available virtual environments by <strong><code>conda env list</code></strong>.</li>
<li><strong>Activate</strong> the environment with name <strong>myenv</strong>, <strong><code>conda activate ~/.conda/envs/myenv</code></strong>.</li>
<li><strong>Install new libraries</strong> after activating your virtual environment: <strong><code>conda install -n myenv [package]</code></strong>.</li>
<li> <strong>Deactivate</strong> your environment when you are done, <strong><code>conda deactivate</code></strong></li>
<li><strong>Delete</strong> virtual environment when you no longer need to use it, <strong><code>conda remove -n myenv -all</code></strong> </li>
<li>Get help with <code>conda -h</code>.</li>
</ol></li>
</ul>

<h3 id="toc_4">PLINK</h3>

<p><strong>PLINK</strong> is a powerful and useful genetic data analysis tool to convert data format, QC, calculate MAF and HWE p-value, calculate kinship matrix, calculate top Principal Components, conduct single variant genetic association studies. Common data format is <code>BED/BIM/FAM</code>.</p>

<ul>
<li>Load <strong>plink</strong> module, e.g., <strong><code>spack load plink2@2.00a4.3</code></strong></li>
<li>Quick command to look at plink input arguments: <code>plink2 -h</code>.</li>
<li>See <a href="https://www.cog-genomics.org/plink/2.0/">PLINK 2 Website</a>.</li>
<li>Unload plink module by <code>spack unload plink2@2.00a4.3</code>.</li>
</ul>

<h3 id="toc_5">BCFTools</h3>

<p><strong>BCFTools</strong> is a fast tool to manipulate <code>sorted/bgzipped/tabixed</code> VCF files of genotype data.</p>

<ul>
<li>Load Tabix module, <strong><code>spack load tabix@2013-12-16</code></strong>.</li>
<li>Load bcftools module, <strong><code>spack load bcftools@1.19</code></strong>.</li>
<li>Quick look at bcftools input arguments, <strong><code>bcftools -h</code></strong>.</li>
<li>See <a href="https://samtools.github.io/bcftools/bcftools.html">BCFTools Manual</a>.</li>
</ul>

<h3 id="toc_6">BEDTools</h3>

<p><strong>BEDTools</strong> utilities are a Swiss-army knife of tools allows one to <strong>intersect, merge, count, complement, and shuffle genomic intervals</strong> from multiple files in widely-used genomic file formats such as <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5.1">BAM</a>, <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a>, <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a>/<a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a>, <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format10.1">VCF</a>. While each individual tool is designed to do a relatively simple task (e.g., intersect two interval files), quite sophisticated analyses can be conducted by combining multiple bedtools operations on the UNIX command line.</p>

<ul>
<li>Load bedtools module, e.g., <strong><code>spack load bedtools2@2.31.1</code></strong>.</li>
<li>Quick look at bedtools input arguments, <strong><code>bedtools -h</code></strong>.</li>
<li>Bedtools can be used to intersect one SNP file and one annotation file to annotate SNPs with respect to genetic features.</li>
<li>See <a href="https://bedtools.readthedocs.io/en/latest/">BEDTools Manual</a></li>
</ul>

<h2 id="toc_7">Submit Jobs by <code>sbatch</code>.</h2>

<h3 id="toc_8">Create a bash script and submit a single job</h3>

<p>Create an example bash script by <code>touch star_genome_create.sh</code> and write the following commands into this file (see example shell script in <code>/home/jyang51/yangfss2/public/ExampleScripts/star_genome_create.sh</code> on HGCC):</p>

<div><pre><code class="language-none">#!/bin/bash
#SBATCH --job-name=STAR_genomeGenerate
#SBATCH --nodes=1
#SBATCH --mem=128G ## request memory
#SBATCH --cpus-per-task=16 ## request cpu/cores 
#SBATCH --array=1 ## a single job; or remove this line of specifying an array job
#SBATCH --time=24:00:00 ## specify job running time for 24 hrs
#SBATCH --output=./SLURM_OUT/%x_%A_%a.out ## save slurm output 
#SBATCH --error=./SLURM_OUT/%x_%A_%a.err ## save slurm errors

spack load miniconda3@24.3.0
conda activate ~/.conda/envs/myenv

### Create genome index
echo Running STAR genomeGenerate ...
STAR --runThreadN 16 --runMode genomeGenerate \
--genomeDir /home/jyang51/yangfss2/public/GenomeIndex/star_indexes/hg38 \
--genomeFastaFiles /home/jyang51/yangfss2/public/GenomeIndex/iGenome/hg38_2021/genome.fa \
--sjdbGTFfile /home/jyang51/yangfss2/public/GenomeIndex/iGenome/hg38_2021/gencode.v46.basic.annotation.gtf \
--sjdbOverhang 150

conda deactivate

exit</code></pre></div>

<ol>
<li>This example bash script will generate genome index of human genome of hg38 assembly for running the alignment tool <code>STAR</code> for mapping RNAseq data. </li>
<li>Use command <code>echo</code> to print out log messages or variable contents to debug and check the job status.</li>
<li> Make this shell script executable: <code>chmod 755 phase.sh</code>.</li>
<li> Submit a job to run this shell script: 
<code>
sbatch /home/jyang51/yangfss2/public/ExampleScripts/star_genome_create.sh
</code></li>
</ol>

<h4 id="toc_9">Remarks about using the <code>/scratch</code> space for I/O intensive jobs</h4>

<p>If one want to use the <code>/scratch</code> space to improve computation efficiency by avoiding heavy I/O, the bash script need to be updated to include commands creating a temporary directory <code>${TMPDIR}</code> under the <code>/scratch</code> space (84TB shared). </p>

<div><pre><code class="language-none">#!/bin/bash

# Generate tmp directory name
TMP_NAME=`/usr/bin/mktemp -u XXXXXXXX`

# Create tmp directory
# TMPDIR=&quot;/scratch/${SLURM_JOB_ID}_${TMP_NAME}&quot; # include slurm job id in the name
TMPDIR=&quot;/scratch/${TMP_NAME}&quot;
echo $TMPDIR
mkdir -p &quot;$TMPDIR&quot;

# Copy input data files into the temporary directory
rsync /home/jyang51/yangfss2/public/ExampleData/Sample_ID.txt ${TMPDIR}/

# Run the following command under the temporary directory
cd ${TMPDIR}/
paste Sample_ID.txt Sample_ID.txt  &gt; output_sample_ID_2.txt

# Copy results back to hard disk
rsync ${TMPDIR}/output_sample_ID_2.txt /home/jyang51/yangfss2/public/ExampleData/

# Remove temporary directory
rm -f -r ${TMPDIR}

exit</code></pre></div>

<ol>
<li>Copying data files from your data drive to this temporary directory by <code>rsync</code>.Run the analysis commands with input data files under the temporary directory.</li>
<li>Write results to this temporary directory.</li>
<li>Copy results back to your data drive.</li>
<li>Erase the created temporary directory.</li>
</ol>

<h3 id="toc_10">Submit Array Jobs</h3>

<p><strong>Array jobs</strong> is a convenient way to submit multiple repetitive jobs that only differs by one input variables, e.g., 10000 repetitive simulations, association study for all 20K genome-wide genes. </p>

<ul>
<li>The following example commands to submit an Array job will submit repetitive jobs to run the same bash scripts <code>star_sbatch.sh</code> for <code>52</code> times, with the only difference of <strong>slurm job id <code>$SLURM_ARRAY_TASK_ID</code></strong> for getting the corresponding sample ID.</li>
</ul>

<div><pre><code class="language-none">sbatch /home/jyang51/yangfss2/public/ExampleScripts/star_sbatch.sh /home/jyang51/yangfss2/public/ExampleData</code></pre></div>

<ul>
<li>The shell script <code>star_sbatch.sh</code> has the following contents:</li>
</ul>

<div><pre><code class="language-none">#!/bin/bash
#SBATCH --job-name=MAP_STAR
#SBATCH --nodes=1
#SBATCH --mem=64G
#SBATCH --cpus-per-task=8
#SBATCH --array=1-52
#SBATCH --time=24:00:00
#SBATCH --output=./SLURM_OUT/%x_%A_%a.out
#SBATCH --error=./SLURM_OUT/%x_%A_%a.err

#### Print the task id.
echo &quot;My SLURM_ARRAY_TASK_ID: &quot; $SLURM_ARRAY_TASK_ID

output_dir=$1
echo &quot;Output file directory: $output_dir&quot;.

##
data_dir=/home/jyang51/yangfss2/projects/Bill_Li_RNAseq/BC_Biomarker_Normal_Biopsy/LBI13454-118133

# Generate tmp directory name
TMP_NAME=`/usr/bin/mktemp -u XXXXXXXX`

# Create tmp directory
# TMPDIR=&quot;/scratch/${SLURM_JOB_ID}_${TMP_NAME}&quot; # include slurm job id in the name
TMPDIR=&quot;/scratch/${TMP_NAME}&quot;
echo $TMPDIR
mkdir -p &quot;$TMPDIR&quot;
cd $TMPDIR

## Determine sample ID
sample=$(head -n ${SLURM_ARRAY_TASK_ID} /home/jyang51/yangfss2/public/ExampleData/Sample_ID.txt | tail -n1)

## Copy raw fastq files to temp idrectory under /scratch
rsync ${data_dir}/RawData/H7JNJDSXC_s1_1_SM_${sample}.fastq.gz ${TMPDIR}/
rsync ${data_dir}/RawData/H7JNJDSXC_s1_2_SM_${sample}.fastq.gz ${TMPDIR}/

###### Scripts that will be run for 52 times to Map 52 samples
## The sample ID will be determined with the given Array_Task_ID from 1 to 52;
## Three input variables are taken by the star_map.sh script;
### Use STAR 2.7.11a
spack load miniconda3@24.3.0
conda activate ~/.conda/envs/myenv

## create temp directory to save map output files
mkdir -p ${TMPDIR}/MAP_OUT/

# Allow use 64GB memory
STAR --genomeDir /home/jyang51/yangfss2/public/GenomeIndex/star_indexes/hg38/ \
--runThreadN ${n_threads} \
--limitBAMsortRAM  64000000000 --genomeLoad NoSharedMemory \
--readFilesIn ${TMPDIR}/H7JNJDSXC_s1_1_SM_${sample}.fastq.gz  ${TMPDIR}/H7JNJDSXC_s1_2_SM_${sample}.fastq.gz  \
--outFileNamePrefix ${TMPDIR}/MAP_OUT/${sample}. \
--readFilesCommand zcat \
--sjdbInsertSave All \
--quantMode TranscriptomeSAM GeneCounts \
--outSAMtype BAM SortedByCoordinate \
--outSAMunmapped None \
--outSAMattributes Standard

## Copy output files to the output directory
rsync ${TMPDIR}/MAP_OUT/ ${output_dir}/

conda deactivate

exit</code></pre></div>

<ul>
<li>The example bash script is written to select the corresponding sample ID from <strong><code>/home/jyang51/yangfss2/public/ExampleData/Sample_ID.txt</code></strong> by the <strong>slurm job id <code>$SLURM_ARRAY_TASK_ID</code></strong>.</li>
<li>Copy raw fastq data to temporary directry under <code>/scratch/</code>.</li>
<li>Write outputs to the temporary directory.</li>
<li>Copy the output files to data drive afterwards, with output file directory given by the first input argument. </li>
<li>Delete temporary directory.</li>
</ul>




</body>

</html>
