---
layout: page
title: "Software"
---

### Software Developed and Maintained by the Yang Lab

<a class="btn btn-primary"
href="https://github.com/yanglab-emory/BGW-TWAS"
role="button">BGW-TWAS</a>

* Bayesian Genome-wide TWAS (BGW-TWAS) method is developed for leveraging both cis- and trans- eQTL through summary statistics for TWAS which is based on the BFGWAS framework. 


<a class="btn btn-primary"
href="https://github.com/yanglab-emory/TIGAR"
role="button">TIGAR</a>

* Transcriptome-Integrated Genetic Association Resource (TIGAR) is developed for integrating gene-expression imputation model training, prediction, and TWAS in the same tool. Efficient handling of VCF genotype files and parallele computing have been implemented in TIGAR. **TIGAR can efficiently train both Elastic-Net and nonparametric Bayesian DPR models for gene expression imputation.**


<a class="btn btn-primary"
href="https://github.com/yjingj/BFGWAS_SS"
role="button">BFGWAS / BFGWAS_SS</a>

* BFGWAS is developed for integrating functional annotation in GWAS by a scalable Bayesian method based on the Bayesian variable selection regression (BVSR) model and Expectation Maximization Monte Carlo Markov Chain (EM-MCMC) algorithm. **BFGWAS can now take summary GWAS data.**


<a class="btn btn-primary"
href="https://github.com/yanglab-emory/BFDA"
role="button">BFDA</a>

* BFDA is a MATLAB package for Bayesian functional data analysis, especially for smoothing multiple functional samples from the same underlying distribution. 





